### PR TITLE
prometheus: Split options listenAddress and port

### PR DESCRIPTION
**Pull request by mkaito**
###### Motivation for this change
Accessing the configured port of a service is quite useful, for example when configuring virtual hosts for a service. The prometheus module did not expose the configured por separately, making it unnecessarily cumbersome to consume.

###### Things done
The option `listenAddress` currently has the format `host:port`. I've split this into two options `listenAddress` for the host, and `port` for the port, and updated the defaults.

This is a breaking change only if you were setting `listenAddress` to a non-standard value. If you were, you should now set `listenAddress` and `port` separately.

Perhaps there should be an assertion that `listenAddress` does not end in `:[0-9]+$` or similar, but I'm honestly not sure how to implement that.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 1]
```diff
@@ -489,9 +489,17 @@ in {
       '';
     };
 
+    port = mkOption {
+      type = types.int;
```
Comment (Ma27) [inline nixos/modules/services/monitoring/prometheus/default.nix:493]: `types.port`

**Comment by Lassulus** [comment]
maybe this could be a good check?
`builtins.match ".*:.*" "127.0.0.1:1234" == null`

**Comment by kirelagin** [comment]
_Full disclosure: I haven’t tested this yet._

**Comment by kirelagin** [comment]
> prometheus, prometheus.passthru.tests on x86_64-linux — Success

Actually, I think, I am going to assume that everything is okay 🙄.

**Comment by lschuermann** [comment]
Unfortunately this breaks my existing working IPv6 setup due to a colon being in the address. Something like this really shouldn't occur on 2020. :cry:

Could someone please fix this?

**Comment by Lassulus** [comment]
I guess this regex could be used then (didn't think about ipv6 last time sadly): `"(.*\\..*):(.*)"` which will match only if the address contains a dot?

**Comment by kirelagin** [comment]
Oh damn, I’m so sorry :/. _Why do IP addresses have to be so complicated..._

I think something like this would to the trick: `"(.*\\..*|\]):(.*)"` – either something with a dot in it (an IPv4 address) or something ending in `]` (an IPv6 address) followed by a colon and a port number.

**Pull request merged**